tests/unit/test_repo.py

```diff
@@ -146,3 +151,292 @@ def test_can_add_repositories_from_string_with_options(self):
+
+    def test_parse_deb822_paragraph_w_comments(self):
+        lines = ubuntu_sources_deb822_with_comments.strip().split("\n")
+        ok_para, bad_para = apt._iter_deb822_stanzas(lines)
+        repos = apt._Deb822Stanza(ok_para).repositories
+        assert len(repos) == 3
+        for repo, suite in zip(repos, ("noble", "noble-updates", "noble-backports")):
+            assert repo.enabled
+            assert repo.repotype == "deb"
+            assert repo.uri == "http://nz.archive.ubuntu.com/ubuntu/"
+            assert repo.release == suite
+            assert repo.groups == ["main", "restricted", "universe", "multiverse"]
+            assert repo.filename == ""
+            assert repo.gpg_key == ""
+        with self.assertRaises(apt.InvalidSourceError):
+            apt._Deb822Stanza(bad_para)
+
+    def test_parse_deb822_lines_ubuntu_sources(self):
+        lines = ubuntu_sources_deb822.strip().split("\n")
+        repos, errors = apt.RepositoryMapping._parse_deb822_lines(lines)
+        print(repos[0].__dict__)
+        assert len(repos) == 4
+        assert not errors
+
+    def test_parse_deb822_lines_w_comments(self):
+        lines = ubuntu_sources_deb822_with_comments.strip().split("\n")
+        repos, errors = apt.RepositoryMapping._parse_deb822_lines(lines)
+        assert len(repos) == 3
+        assert len(errors) == 1
+        [error] = errors
+        assert isinstance(error, apt.InvalidSourceError)
+
+    def test_load_deb822_ubuntu_sources(self):
+        def isnt_file(f: str) -> bool:
+            return False
+
+        def iglob_nothing(s: str) -> typing.Iterable[str]:
+            return []
+
+        with patch("os.path.isfile", new=isnt_file):
+            with patch("glob.iglob", new=iglob_nothing):
+                repository_mapping = apt.RepositoryMapping()
+        assert not repository_mapping._repository_map
+
+        with patch("builtins.open", new_callable=mock_open, read_data=ubuntu_sources_deb822):
+            repository_mapping.load_deb822("")
+        assert sorted(repository_mapping._repository_map.keys()) == [
+            "deb-http://nz.archive.ubuntu.com/ubuntu/-noble",
+            "deb-http://nz.archive.ubuntu.com/ubuntu/-noble-backports",
+            "deb-http://nz.archive.ubuntu.com/ubuntu/-noble-updates",
+            "deb-http://security.ubuntu.com/ubuntu-noble-security",
+        ]
+
+    def test_load_deb822_w_comments(self):
+        def isnt_file(f: str) -> bool:
+            return False
+
+        def iglob_nothing(s: str) -> typing.Iterable[str]:
+            return []
+
+        with patch("os.path.isfile", new=isnt_file):
+            with patch("glob.iglob", new=iglob_nothing):
+                repository_mapping = apt.RepositoryMapping()
+        assert not repository_mapping._repository_map
+
+        with patch(
+            "builtins.open", new_callable=mock_open, read_data=ubuntu_sources_deb822_with_comments
```

Comment:
Similar here -- instead of patching `open`, can we just write a local file, which is fast and reliable?

Reply:
If you're trying to avoid destructive tests - e.g. non-isolated unit tests that will modify your host in some way - [`pyfakefs`](https://pypi.org/project/pyfakefs/) is quite good for mocking common filesystem operations such as read and writing files.